### PR TITLE
rbd attach/detach refactoring

### PR DESCRIPTION
**Pull request by rootfs**
**What this PR does / why we need it**:
RBD volume implements attach/detach interface to resolve rbd image locking.

`Attach()`: controller locks the rbd image
`Detach()`: controller removes the image lock
`WaitForAttach()`: node maps rbd image
`MountDevice()`: node mounts rbd device

**Which issue this PR fixes** _(optional, in `fixes #<issue number>(, #<issue_number>, ...)` format, will close that issue when PR gets merged)_: fixes #
https://github.com/openshift/origin/issues/7983
**Special notes for your reviewer**:
There is also a short version to commit https://github.com/kubernetes/kubernetes/commit/4382260c99ae9c0d83d7fae8e3a0de5f7b2656a5. Instead of refactoring `NewFromVolume()`, we can add a new volume API `NewFromVolumeWithNameSpace()` and avoid massive refactoring. I am open to either implementation.

@kubernetes/sig-storage @elsonrodriguez @spinolacastro
**Release note**:

``` release-note
```

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33660)

### CODE REVIEW

**Comment by jsafrane** [comment]
Reviewed 12 of 12 files at r1, 44 of 44 files at r2, 7 of 8 files at r3.
Review status: 63 of 64 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/controller/volume/persistentvolume/controller.go, line 983 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33660#-KSp174ZU2Qn9IYr0WEz:-KSp174ZU2Qn9IYr0WF-:bzbt4n) ([raw file](https://github.com/kubernetes/kubernetes/blob/63044d1b3ffbd36f0255e2a40523a2c861f2d076/pkg/controller/volume/persistentvolume/controller.go#L983)):*
> ```Go
> 
> 	// Find a plugin.
> 	// no namespace to passed into VolumeSpec. This is fine because recyclers (NFS, hostPath) don't use namespace'd resource yet
> ```

at this point, pv.Spec.ClaimRef still points to the PVC that used it and you can find namespace there

---

*[pkg/controller/volume/persistentvolume/controller.go, line 1163 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33660#-KSp1HDGSoNk5D4xgDd8:-KSp1HDHjAoUmzPcDmoK:bqns8db) ([raw file](https://github.com/kubernetes/kubernetes/blob/63044d1b3ffbd36f0255e2a40523a2c861f2d076/pkg/controller/volume/persistentvolume/controller.go#L1163)):*
> ```Go
> 	// Plugin found
> 	glog.V(5).Infof("found a deleter plugin %q for volume %q", plugin.GetPluginName(), volume.Name)
> 	// no namespace passed into VolumeSpec. Deleters are expect to find namespace'd resource from e.g. annotations
> ```

at this point, pv.Spec.ClaimRef should point to the PVC that used it and you can find namespace there.

---

*[pkg/volume/plugins.go, line 241 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33660#-KSp0rYDtu2S1RhDUR-M:-KSp0rYE0WlqXgkFfyHI:b-63jgnu) ([raw file](https://github.com/kubernetes/kubernetes/blob/63044d1b3ffbd36f0255e2a40523a2c861f2d076/pkg/volume/plugins.go#L241)):*
> ```Go
> 	PersistentVolume *api.PersistentVolume
> 	ReadOnly         bool
> 	NameSpace        string
> ```

Please add a comment what namespace is this (pod or pvc that uses the volume) and why it's needed.

---

*[pkg/volume/rbd/attacher.go, line 59 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33660#-KSp9pt6QEQ9bCaCzVpz:-KSp9pt6QEQ9bCaCzVq-:b-183iif) ([raw file](https://github.com/kubernetes/kubernetes/blob/63044d1b3ffbd36f0255e2a40523a2c861f2d076/pkg/volume/rbd/attacher.go#L59)):*
> ```Go
> }
> 
> func (attacher *rbdAttacher) Attach(spec *volume.Spec, hostName string) (string, error) {
> ```

please add some comment that it does not attach the volume and locks it instead.

---

*[pkg/volume/rbd/attacher.go, line 75 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33660#-KSp9ldOHVl8zU543WJq:-KSp9ldOHVl8zU543WJr:b-d4yzgy) ([raw file](https://github.com/kubernetes/kubernetes/blob/63044d1b3ffbd36f0255e2a40523a2c861f2d076/pkg/volume/rbd/attacher.go#L75)):*
> ```Go
> 		return "", err
> 	}
> 	return attacher.manager.AttachDisk(*mounter)
> ```

should it wait for the lock acquired by Attach()? Is it even possible to check for the lock here? Or can we be sure that the lock is held at this point? (i am not familiar with attach controller logic here).

---

*[pkg/volume/rbd/rbd_lock.go, line 69 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33660#-KSpEnhygJg0FUeZk0Cw:-KSpEnhygJg0FUeZk0Cx:b-igswmi) ([raw file](https://github.com/kubernetes/kubernetes/blob/63044d1b3ffbd36f0255e2a40523a2c861f2d076/pkg/volume/rbd/rbd_lock.go#L69)):*
> ```Go
> 
> 		if err != nil {
> 			continue
> ```

should we log (with a lower log level) individual errors?

---

*[pkg/volume/rbd/rbd_lock.go, line 99 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33660#-KSpEvTta6ZsZzCDR_c8:-KSpEvTta6ZsZzCDR_c9:b-ynz1t4) ([raw file](https://github.com/kubernetes/kubernetes/blob/63044d1b3ffbd36f0255e2a40523a2c861f2d076/pkg/volume/rbd/rbd_lock.go#L99)):*
> ```Go
> 			//lock is acquired
> 			break
> 		}
> ```

should we log (with a lower log level) individual errors and cmd output?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33660)*

**Comment by jsafrane** [comment]
Reviewed 1 of 8 files at r3.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33660)*

**Comment by elsonrodriguez** [comment]
@rootfs It looks like there's a mismatch between the path the kublet mounts the RBD volume at, and the path passed into Docker for the bind mount.

The host path for the RBD volume in this case is:
/var/lib/kubelet/plugins/kubernetes.io/rbd/rbd/kube-image-kubernetes-dynamic-pvc-6dc371e7-8a5f-11e6-9463-001e677fb1d7

But Docker is being told to bindmount this:

/var/lib/kubelet/pods/76744c76-8a67-11e6-9a10-001e677fb1d7/volumes/kubernetes.io~rbd/pvc-6dc28fcc-8a5f-11e6-9a10-001e677fb1d7

This causes the RBD volume to not show up in the container.

**Comment by elsonrodriguez** [comment]
@rootfs Also this appears to be working as far as unlocking. About 10 minutes after a pod is rescheduled from a downed node, the volumes are made available to the new node.

**Comment by rootfs** [comment]
@elsonrodriguez thank you for testing. The container mount path problem should be fixed by now.

**Comment by elsonrodriguez** [comment]
@rootfs Previous commit fixed the path mismatch!

Gonna try things with the rebase.

**Comment by rootfs** [comment]
@saad-ali @jingxu97 please help review changes in Detach(), operation executor, and DeviceOpened(). Thanks!

**Comment by rootfs** [comment]
Review status: 0 of 27 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/volume/rbd/attacher.go, line 75 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33660#-KSp9ldOHVl8zU543WJq:-KUHvlvEJBqcI5TYT_LP:b-5fl2w6) ([raw file](https://github.com/kubernetes/kubernetes/blob/63044d1b3ffbd36f0255e2a40523a2c861f2d076/pkg/volume/rbd/attacher.go#L75)):*
<details><summary><i>Previously, jsafrane (Jan Šafránek) wrote…</i></summary>
> should it wait for the lock acquired by Attach()? Is it even possible to check for the lock here? Or can we be sure that the lock is held at this point? (i am not familiar with attach controller logic here).
</details>
comment added

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33660)*

**Comment by rootfs** [comment]
Review status: 0 of 27 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/volume/plugins.go, line 241 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33660#-KSp0rYDtu2S1RhDUR-M:-KUHx8wbL3IxDfGBQkFT:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/63044d1b3ffbd36f0255e2a40523a2c861f2d076/pkg/volume/plugins.go#L241)):*
<details><summary><i>Previously, jsafrane (Jan Šafránek) wrote…</i></summary>
> Please add a comment what namespace is this (pod or pvc that uses the volume) and why it's needed.
</details>
Done.

---

*[pkg/volume/rbd/attacher.go, line 59 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33660#-KSp9pt6QEQ9bCaCzVpz:-KUHxAy_i6Gi9djKQnyV:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/63044d1b3ffbd36f0255e2a40523a2c861f2d076/pkg/volume/rbd/attacher.go#L59)):*
<details><summary><i>Previously, jsafrane (Jan Šafránek) wrote…</i></summary>
> please add some comment that it does not attach the volume and locks it instead.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33660)*

**Comment by rootfs** [comment]
Review status: 0 of 27 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/volume/rbd/rbd_lock.go, line 69 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33660#-KSpEnhygJg0FUeZk0Cw:-KUHxG4JrWxGzUWJgLFQ:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/63044d1b3ffbd36f0255e2a40523a2c861f2d076/pkg/volume/rbd/rbd_lock.go#L69)):*
<details><summary><i>Previously, jsafrane (Jan Šafránek) wrote…</i></summary>
> should we log (with a lower log level) individual errors?
</details>
Done.

---

*[pkg/volume/rbd/rbd_lock.go, line 99 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33660#-KSpEvTta6ZsZzCDR_c8:-KUHxDWUlCvib2mftlM8:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/63044d1b3ffbd36f0255e2a40523a2c861f2d076/pkg/volume/rbd/rbd_lock.go#L99)):*
<details><summary><i>Previously, jsafrane (Jan Šafránek) wrote…</i></summary>
> should we log (with a lower log level) individual errors and cmd output?
</details>
Done.

---

*[pkg/controller/volume/persistentvolume/controller.go, line 983 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33660#-KSp174ZU2Qn9IYr0WEz:-KUHxElr11ygRPCH7vbz:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/63044d1b3ffbd36f0255e2a40523a2c861f2d076/pkg/controller/volume/persistentvolume/controller.go#L983)):*
<details><summary><i>Previously, jsafrane (Jan Šafránek) wrote…</i></summary>
> at this point, pv.Spec.ClaimRef still points to the PVC that used it and you can find namespace there
</details>
Done.

---

*[pkg/controller/volume/persistentvolume/controller.go, line 1163 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33660#-KSp1HDGSoNk5D4xgDd8:-KUHxF26VBm55F4PaPEW:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/63044d1b3ffbd36f0255e2a40523a2c861f2d076/pkg/controller/volume/persistentvolume/controller.go#L1163)):*
<details><summary><i>Previously, jsafrane (Jan Šafránek) wrote…</i></summary>
> at this point, pv.Spec.ClaimRef should point to the PVC that used it and you can find namespace there.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33660)*

**Comment by rootfs** [comment]
Review status: 0 of 27 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/volume/rbd/rbd_lock.go, line 99 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33660#-KSpEvTta6ZsZzCDR_c8:-KUHxIQRvEInQjKM9iy5:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/63044d1b3ffbd36f0255e2a40523a2c861f2d076/pkg/volume/rbd/rbd_lock.go#L99)):*
<details><summary><i>Previously, rootfs (Huamin Chen) wrote…</i></summary>
> Done.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/33660)*

**Comment by rootfs** [comment]
@kubernetes/sig-storage PTAL, need it for 1.5, thanks!

File: examples/experimental/persistent-volume-provisioning/rbd/rbd-storage-class.yaml [hunk 1]
```diff
@@ -4,7 +4,7 @@ metadata:
    name: slow
 provisioner: kubernetes.io/rbd
 parameters:
-    monitors: 127.0.0.1:6789
+    monitors: 10.16.153.105:6789
```
Comment (elsonrodriguez) [inline examples/experimental/persistent-volume-provisioning/rbd/rbd-storage-class.yaml]: Could we keep 127.0.0.1? This makes it consistent with the other examples.

**Comment by elsonrodriguez** [comment]
@rootfs Did some quick smoke tests, still looking good as far as provisioning, mounting, and unlocking.

I did notice that when deleting a pod, a full 10 minutes must pass before the lock is removed. Before this was much faster. The 10 minute delay is fine for a downed node situation, but for simple scaling/rescheduling will be problematic.

Would it be possible to instantly unlock for pod deletion events not triggered by NodeNotReady ejections?

If such a separation isn't possible (and as a nice to have regardless) would it be possible to make the unlock wait period configureable?

**Comment by rootfs** [comment]
opened separate PRs #35433 #35434 to track volume and mount changes.

File: pkg/volume/rbd/rbd.go [hunk 2]
```diff
@@ -112,18 +115,12 @@ func (plugin *rbdPlugin) NewMounter(spec *volume.Spec, pod *api.Pod, _ volume.Vo
...
-	// rbd volumes used directly in a pod have a ReadOnly flag set by the pod author.
-	// rbd volumes used as a PersistentVolume gets the ReadOnly flag indirectly through the persistent-claim volume used to mount the PV
-	if spec.Volume != nil && spec.Volume.RBD != nil {
-		return spec.Volume.RBD, spec.Volume.RBD.ReadOnly
-	} else {
-		return spec.PersistentVolume.Spec.RBD, spec.ReadOnly
+func (plugin *rbdPlugin) newMounterInternal(spec *volume.Spec, podUID types.UID, manager diskManager, mounter mount.Interface, secret string) (volume.Mounter, error) {
+	source, readOnly, err := getVolumeSource(spec)
+	if err != nil {
+		return nil, err
```
Comment (rootfs) [inline pkg/volume/rbd/rbd.go]: git diff is a bit screwed. `getVolumeSource` fails if it gets an invalid spec. This function returns source and readOnly, which are used by rbd mounter.

**Comment by rootfs** [comment]
@jingxu97 @elsonrodriguez feedback addressed. The volume changes are removed from this PR.

@pmorie PTAL, thanks!

**Comment by elsonrodriguez** [comment]
@rootfs What was the resolution for the detach timeout?

**Comment by rootfs** [comment]
@elsonrodriguez On a local cluster, rbd is unmapped and unlocked soon after pod is deleted.

**Comment by elsonrodriguez** [comment]
@rootfs 

Here's a walkthrough of what i'm observing.

First I delete a pod that has an RBD volume:
```
$ kubectl delete pod hadoop-namenode-2955366752-0850o  -nhadoop
```

New pod has spun up, old pod terminating. Eventually the pod does terminate.
```
$ kubectl get pods -o wide -nhadoop
NAME                               READY     STATUS              RESTARTS   AGE       IP           NODE
hadoop-datanode-423441758-hiiy9    1/1       Running             0          18h       10.9.220.9   node-2e148c8c.harbor0.group.company.com
hadoop-httpfs-2431655975-ysy1c     1/1       Running             0          4d        10.9.218.4   node-63ec280a.harbor0.group.company.com
hadoop-namenode-2955366752-0850o   1/1       Terminating         0          21h       10.9.212.6   node-230cd389.harbor0.group.company.com
hadoop-namenode-2955366752-19pt5   0/1       ContainerCreating   0          36s       <none>       node-63ec280a.harbor0.group.company.com
```

Getting the RBD volume info:
```
$ kubectl get pv
NAME                                       CAPACITY   ACCESSMODES   RECLAIMPOLICY   STATUS    CLAIM                              REASON    AGE
pvc-076d089a-9ae1-11e6-a956-001e677fb1d7   10Gi       RWO           Delete          Bound     hadoop/hadoop-namenode                       21h

$ kubectl describe pv pvc-076d089a-9ae1-11e6-a956-001e677fb1d7
Name:		pvc-076d089a-9ae1-11e6-a956-001e677fb1d7
Labels:		<none>
Status:		Bound
Claim:		hadoop/hadoop-namenode
Reclaim Policy:	Delete
Access Modes:	RWO
Capacity:	10Gi
Message:	
Source:
    Type:		RBD (a Rados Block Device mount on the host that shares a pod's lifetime)
    CephMonitors:	[ceph-mon.ceph.svc.harbor0.group.company.com:6789]
    RBDImage:		kubernetes-dynamic-pvc-076de11b-9ae1-11e6-8126-001e677fb1d7
    FSType:		
    RBDPool:		kube
    RadosUser:		kube
    Keyring:		/etc/ceph/keyring
    SecretRef:		&{ceph-secret-user}
    ReadOnly:		false
No events.
```

Getting the locks on the image:
```
root@ceph-mon-1396930634-0vjdx:/# rbd lock list kube/kubernetes-dynamic-pvc-076de11b-9ae1-11e6-8126-001e677fb1d7
There is 1 exclusive lock on this image.
Locker        ID                                                         Address                  
client.176546 kubelet_lock_magic_node-230cd389.harbor0.group.company.com 10.54.9.214:0/2351140988 
```

There is still a lock, which prevents the new node from mapping the volume. Eventually it succeeds, but only after 10 minutes:
```
  FirstSeen	LastSeen	Count	From							SubObjectPath			Type		Reason		Message
  ---------	--------	-----	----							-------------			--------	------		-------
  13m		13m		1	{default-scheduler }									Normal		Scheduled	Successfully assigned hadoop-namenode-2955366752-19pt5 to node-63ec280a.harbor0.group.company.com
  13m		7m		11	{controller-manager }									Warning		FailedMount	Failed to attach volume "pvc-076d089a-9ae1-11e6-a956-001e677fb1d7" on node "node-63ec280a.harbor0.group.company.com" with: exit status 16
  11m		5m		4	{kubelet node-63ec280a.harbor0.group.company.com}					Warning		FailedMount	Unable to mount volumes for pod "hadoop-namenode-2955366752-19pt5_hadoop(2e255f22-9b94-11e6-a956-001e677fb1d7)": timeout expired waiting for volumes to attach/mount for pod "hadoop-namenode-2955366752-19pt5"/"hadoop". list of unattached/unmounted volumes=[hadoop-namenode]
  11m		5m		4	{kubelet node-63ec280a.harbor0.group.company.com}					Warning		FailedSync	Error syncing pod, skipping: timeout expired waiting for volumes to attach/mount for pod "hadoop-namenode-2955366752-19pt5"/"hadoop". list of unattached/unmounted volumes=[hadoop-namenode]
  3m		3m		1	{kubelet node-63ec280a.harbor0.group.company.com}	spec.containers{namenode}	Normal		Pulled		Container image "uhopper/hadoop-namenode:2.7.2" already present on machine
  3m		3m		1	{kubelet node-63ec280a.harbor0.group.company.com}	spec.containers{namenode}	Normal		Created		Created container with docker id e58280a5fada; Security:[seccomp=unconfined]
  3m		3m		1	{kubelet node-63ec280a.harbor0.group.company.com}	spec.containers{namenode}	Normal		Started		Started container with docker id e58280a5fada
```

However if I pin the pod to the node it is already on:

```
$ kubectl get pods -nhadoop -o wide
NAME                               READY     STATUS        RESTARTS   AGE       IP           NODE
hadoop-datanode-423441758-hiiy9    1/1       Running       1          18h       10.9.220.9   node-2e148c8c.harbor0.group.company.com
hadoop-httpfs-2431655975-ysy1c     1/1       Running       0          4d        10.9.218.4   node-63ec280a.harbor0.group.company.com
hadoop-namenode-2955366752-19pt5   1/1       Terminating   0          23m       10.9.218.5   node-63ec280a.harbor0.group.company.com
hadoop-namenode-4175910539-wfd74   1/1       Running       0          21s       10.9.218.6   node-63ec280a.harbor0.group.company.com
```

So it looks like if the pod is scheduled ot a new node, there is a mandatory 10 minute wait period until it is detached. If the pod isn't scheduled to a new node (in the case of pinning or single node cluster) then the lock doesn't come into play.

**Comment by rootfs** [comment]
@saad-ali @jingxu97 any input on the 10 minutes detach wait time?

**Comment by childsb** [comment]
This PR is blocked on  #35433 #35434

**Comment by jsafrane** [comment]
@rootfs, I am quite satisfied with the code quality (if you promise to fill `Spec.Namespace` where it is possible, e.g. before calling `NewMounter`), but I don't know if attach/detach controller can always reconstruct full `Spec` in some catastrophic cases (e.g controller restart). @jingxu97, @saad-ali?

File: pkg/volume/volume.go [hunk 3]
```diff
@@ -165,7 +165,7 @@ type Attacher interface {
 // Detacher can detach a volume from a node.
 type Detacher interface {
 	// Detach the given device from the node with the given Name.
-	Detach(deviceName string, nodeName types.NodeName) error
+	Detach(deviceName string, nodeName types.NodeName, spec *Spec) error
```
Comment (Quentin-M) [inline pkg/volume/volume.go]: To match how the Attacher interface looks like, should the *Spec be the first argument?

Comment (rootfs) [inline pkg/volume/volume.go]: ack

**Comment by rootfs** [comment]
need #36179 to pass RBD e2e tests

**Comment by saad-ali** [comment]
> @saad-ali @jingxu97 please help review changes in Detach(), operation executor, and DeviceOpened(). Thanks!

Requiring volume.Spec in Detach does make restart reconstruction scenarios more challenging to handle, but it looks like you need it since a secret is required for your to get and release your lock.

File: pkg/volume/plugins.go [hunk 4]
```diff
@@ -317,6 +318,23 @@ func NewSpecFromPersistentVolume(pv *api.PersistentVolume, readOnly bool) *Spec
 	}
 }
 
+// NewSpecFromVolumeWithNS creates an Spec from an api.Volume and a namespace
+func NewSpecFromVolumeWithNS(vs *api.Volume, nameSpace string) *Spec {
```
Comment (saad-ali) [inline pkg/volume/plugins.go]: This is really dirty: extending an existing object and then creating a separate constructor that initializes that field.

**Review by saad-ali — CHANGES_REQUESTED** [review]
Non-rbd changes look mostly fine. I'm not happy with having spec passed to Detach, but I don't see a good way around it.

File: pkg/volume/plugins.go (hunk 4 above)
Comment (rootfs) [inline pkg/volume/plugins.go]: To pass namespace to `volumeSpec`, we either refactor 1) `NewSpecFromVolume` to add a second parameter `namespace` or 2) create a new constructor (as in here)

As I stated in the PR, I tried both ways. The sad story of refactoring `NewSecFromVolume` is that it creates a massive refactoring to many innocent files. Thus I believe a new constructor is a cleaner way for the job: it is only used in the place where namespace is needed.

**Pull request closed without merging**